doc/persistence.adoc

```diff
@@ -0,0 +1,165 @@
+
+=== Lightweight database library
+
+Requires an insecure environment and an installation of the database library that is accessible to Minetest. SQLite3, available through the `lsqlite3` luarocks package, is a popular choice here and used for instance by the https://github.com/shivajiva101/sban[sban] mod.
+
+=== String stores
+
+==== Entity staticdata
+
+Tied to entities. The serialized string must be returned by `get_staticdata` and is passed to `on_activate`.
+
+==== File store
+
+Usually tied to world or mod paths. The simplest approach reads the file at load time and writes it on shutdown. As `on_shutdown` may however not be called in the case of a crash - or even worse, a power outage might abruptly shut down the server without calling anything - this provides a rather poor granularity, as all changes to the data during the uptime may be lost.
+
+You may simply serialize your data and write it to a file on every update. If your data is rather larger or gets updated frequently, a full serialization might negatively impact performance.
+Performance can be improved at the expense of granularity by saving periodically and choosing "long" periods.
+A transaction log improves performance by only storing changes, at the expense of disk space.
+
+TIP: A mix of both approaches can provide satisfying results, logging only changes and rewriting the logfile periodically to keep disk space waste acceptable.
+
+For special cases like logging, an append-only file may be the ideal solution if using the global `minetest.log` is not desirable.
+
+=== Key-value store
+
+==== Filesystem
+
+On systems that provide a decent filesystem implementation (that is, everything except Windows), you can use filenames/filepaths as keys and files as values. On poor filesystems, you might be heavily limited by absolute path character limits; lots of small files might lead to fragmentation. A nested hierarchical key-value store is possible through directory structures, which can be managed and traversed using:
+* `minetest.mkdir`
+* `minetest.rmdir`
+* `minetest.cpdir`
+* `minetest.mvdir`
+* `minetest.get_dir_list`
+
+If you want to mitigate the risk of data loss, you can use `minetest.safe_file_write` when rewriting files.
+// TODO document FS helpers thoroughly elsewhere
+
+==== Configuration files
+
+The `Settings` object allows you to operate on configuration files, getting & setting key-value entries and saving the file. The main `Settings` object `minetest.settings` can be used to persist a few settings "globally" - bleeding everywhere. This is horribly abused by the mainmenu to store stuff like the last selected game. Don't be like the mainmenu. Configuration files are presumably easy to edit for users, but so is Lua.
```

Comment:
This feels a little _too_ opinionated to me. I think that the configuration interface is fine if used properly, and is likely a lot faster for the simple k/v storage it's supposed to be than serialization. Noting that the `minetest.settings` object is global is good, even separating that part out into something like `CAUTION: This sets global settings, don't set global settings.` might be good.